core/src/main/java/org/apache/accumulo/core/logging/FateLogger.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.accumulo.core.logging;
+
+import static org.apache.accumulo.fate.FateTxId.formatTid;
+
+import java.io.Serializable;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Function;
+
+import org.apache.accumulo.fate.ReadOnlyRepo;
+import org.apache.accumulo.fate.Repo;
+import org.apache.accumulo.fate.StackOverflowException;
+import org.apache.accumulo.fate.TStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FateLogger {
+  private static final String PREFIX = Logging.PREFIX + "fate.";
+
+  // Logs all mutations to FATEs persistent storage. Enabling this logger could help debug
+  // reproducible problems with FATE transactions.
+  private static final Logger storeLog = LoggerFactory.getLogger(PREFIX + "store");
+
+  public static <T> TStore<T> wrap(TStore<T> store, Function<Repo<T>,String> toLogString) {
+
+    // only logging operations that change the persisted data, not operations that only read data
+    return new TStore<T>() {
+
+      @Override
+      public long reserve() {
+        return store.reserve();
+      }
+
+      @Override
+      public void reserve(long tid) {
+        store.reserve(tid);
+      }
+
+      @Override
+      public void unreserve(long tid, long deferTime) {
+        store.unreserve(tid, deferTime);
+      }
+
+      @Override
+      public List<ReadOnlyRepo<T>> getStack(long tid) {
+        return store.getStack(tid);
+      }
+
+      @Override
+      public TStatus getStatus(long tid) {
+        return store.getStatus(tid);
+      }
+
+      @Override
+      public TStatus waitForStatusChange(long tid, EnumSet<TStatus> expected) {
+        return store.waitForStatusChange(tid, expected);
```

Comment:
I see your comments about avoiding logging on reads; however, I do feel that when a status changes it would be nice to see the update and the receive on that code path. Therefore when a status change occurs, it *may* be nice to have a log statement.

Reply:
If write logging is enabled, then when the status changes that will be seen.

I considered having a separate prefix for read related log messages so that it could be turned on and off independently from write log messages.  However I could not think of use cases for the read logging that was not covered by the write logging.  One use case I thought of was when a thread is stuck reading something, but I think jstack and memory dumps may be better tools for that use case.